packages: change systemd boot sequence

arnaldo2792 opened the pull request:
**Issue number:**
#1392

**Description of changes:**

```
packages: change systemd boot sequence
```

The current systemd boot sequence is error prone on isolated scenarios like sending invalid configurations through user data. These scenarios could cause inconsistent states at the end of the boot process, which affect the mechanism to determine if a boot was successful after applying new configurations or updates.

With the current boot sequence, it is difficult to implement features that require a strict services' initialization order (a.k.a run levels).

In order to fix the problems presented above, this commit defines a new boot sequence, with three main systemd targets: preconfigured, configured and multi-user. The "host-containers" crate was modified to enable the host-container service units only when multi-user is the current default target.

The "mark-successful-boot" service unit was moved from the "multi-user" target to the "preconfigured" target, and it only depends on the "migrator" oneshot service unit since it is the only service required (for now) to determine if a boot was successful. The "mark-successful-boot" service unit used to send metrics after it was executed. However, doing so indirectly declared a strong dependency on "settings-applier" since the unit was setup to use the "proxy.env" file. The "metricdog-successful-boot" oneshot service unit was created to remove such dependency from "mark-successful-boot".

#### Preconfigured
This target is used to start the boot process (a.k.a default target), it is based on the "multi-user" target provided by systemd so that other basic systemd units are started during this stage. Failures in the required "oneshot" service units will cause the target to fail, stopping the boot process. The boot is marked as successful during the execution of this target, only if the migrator "oneshot" service exits successfully. Services initialized during this phase include:

* acpid
* chrony
* dbus-broker
* audit-rules
* migrator (required)
* metricdog-successful-boot
* storewolf (required)
* early-boot-config (required)
* sundog (required)
* settings-applier(required)

Once the target is reached (completed), the "activate-configured.service" unit will set the configured target as the default target and start it.

#### Configured
This target should be used to setup additional configurations in the host before services like kubernetes/ecs start. Services initialized during this phase include:

* chronyd
* host-containerd

Once the target is reached (completed) the "activate-multi-user.service" unit will set the multi-user target as the default target, and start it.

#### Multi-user
This is the final target enabled in the boot sequence. Services initialized during this phase include:

* host-containers@*
* docker
* ecs
* kubernetes

**Testing done:**
* `systemctl status` didn't show any failures on successful boots
* Run nginx pod/task/container
* Admin/Control containers were enabled once the `multi-user` target was reached
* Custom host container sent as user data was enabled once  the `multi-user`  target was reached
* Custom host container created after boot was enabled after the new settings were applied

```sh
apiclient set --json \
    '{"host-containers": { "test": { "source": "docker.io/arnaldo2792/blocker:latest", "superpowered": false, "enabled": true}}}'
```

```sh
ps ax -o pid,args | grep host-ctr
479 /usr/bin/host-ctr run --container-id=admin --source=328549459982.dkr.ecr.us-west-2.amazonaws.com/bottlerocket-admin:v0.6.0 --superpowered=true
480 /usr/bin/host-ctr run --container-id=control --source=328549459982.dkr.ecr.us-west-2.amazonaws.com/bottlerocket-control:v0.4.2 --superpowered=false
481 /usr/bin/host-ctr run --container-id=custom-at-boot --source=docker.io/arnaldo2792/blocker:latest --superpowered=false
16546 /usr/bin/host-ctr run --container-id=test --source=docker.io/arnaldo2792/blocker:latest --superpowered=false
```
* Failed boot with bad configurations

```toml
[settings.host-containers.admin]
enabled = true

[settings.bad]
bad = "data"
```

```sh
[   11.365797] early-boot-config[366]: Error PATCHing '/settings?tx=bottlerocket-launch': Status 400 when PATCHing /settings?tx=bottlerocket-launch: Json deserialize error: unknown field `bad`, expected one of `source`, `enabled`, `superpowered`, `user-data` at line 1 column 85
[FAILED] Failed to start Bottlerocket userdata configuration system.
See 'systemctl status early-boot-config.service' for details.
[DEPEND] Dependency failed for User-specified setting generators.
[DEPEND] Dependency failed for Applies settings to create config files.
[DEPEND] Dependency failed for Bottlerocket dynamic pre-configuration.
[DEPEND] Dependency failed for Isolates configured.target.
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 committed:
refactor system boot sequence

The current systemd boot sequence is error prone on isolated scenarios
like sending invalid configurations through user data. These scenarios
could cause inconsistent states at the end of the boot process, which
affect the mechanism to determine if a boot was successful after
applying new configurations or updates. It is also difficult to
implement features that require service initialization order.

In order to fix the problems presented above, this commit defines a new
boot sequence, with three main systemd targets: preconfigured, configured
and multi-user.

Preconfigured

This target is used to start the boot process. Failures in any of the
required service units will cause the target to fail, stopping the boot
process. The boot is marked as successful during the execution of this
target only if the migrator "oneshot" service exists successfully.
Services initialized during this phase include:

* migrator
* mark-successful-boot
* send-boot-success
* storewolf
* apiserver
* early-boot-config
* sundog
* settings-applier

Once the target is reached (completed), the "activate-configured.service"
unit will set the configured target as the default target and start it.

Configured

This target should be used to setup additional configurations in the
host before services like kubernetes/ecs start. Services initialized
during this phase include:

* chronyd
* host-containerd

Once the target is reached (completed) the "activate-multi-user.service"
unit will set the multi-user target as the default target, and start it.

Multi-user

This is the final target enabled in the boot sequence. Services
initialized during this phase include:

* host-containers@*
* docker
* ecs
* kubernetes

bcressey commented on packages/chrony/chronyd.service:
```diff
@@ -1,12 +1,12 @@
 ExecStart=/usr/sbin/chronyd -d -F -1
 
 [Install]
-WantedBy=multi-user.target
+WantedBy=preconfigured.target
```
This seems like it should be `WantedBy=configured.target` since that's the next one in the chain.

arnaldo2792 replied:
I used `WantedBy=preconfigured.target` because I thought that chrony should is one of the services required at the very early stages of the boot (during the preconfigured target). I'll make the changes to be started in the `configured` target.

bcressey replied:
It'd be fine to put it in the "preconfigured" stage, although a bit tricky since chrony itself needs to be configured by settings. I'm OK with "configured" for now.

bcressey commented on packages/libaudit/audit-rules.service:
```diff
@@ -12,4 +12,4 @@ ExecStart=/sbin/auditctl -R /usr/share/audit/audit.rules
 
 [Install]
-WantedBy=multi-user.target
+RequiredBy=preconfigured.target
```
This can be `WantedBy=preconfigured.target` since even though it's a one-shot, it's not related to settings.

bcressey commented on packages/os/host-containers@.service:
```diff
@@ -1,14 +1,12 @@
 Description=Host container: %i
 After=host-containerd.service
-Wants=host-containerd.service
+Requires=host-containerd.service
```
This would revert e1a051c81 and cause host container units to be stopped whenever host-containerd restarts.

bcressey commented on packages/os/migrator.service:
```diff
@@ -3,9 +3,13 @@ Description=Bottlerocket data store migrator
 [Service]
 Type=oneshot
-ExecStart=/usr/bin/migrator --datastore-path /var/lib/bottlerocket/datastore/current --migration-directory /var/lib/bottlerocket-migrations --root-path /usr/share/updog/root.json --metadata-directory /var/cache/bottlerocket-metadata --migrate-to-version-from-os-release
+ExecStart=/usr/bin/migrator --datastore-path /var/lib/bottlerocket/datastore/current \
```
nit:
```suggestion
ExecStart=/usr/bin/migrator \
  --datastore-path /var/lib/bottlerocket/datastore/current \
```

bcressey commented on packages/os/sundog.service:
```diff
@@ -1,8 +1,8 @@
-Requires=network-online.target apiserver.service
+# Needs network access to support commands talking to IMDS.
+After=network-online.target early-boot-config.service
+Requires=network-online.target early-boot-config.service
```
`sundog` has a direct dependency on the API being up, so it seems like it makes sense to keep the `apiserver.service` requirement.

arnaldo2792 replied:
`early-boot-config` requires `apiserver`, so by just specifying `early-boot-config` we get the exact same behavior as specifying all the removed units.

bcressey replied:
I'd prefer units to more fully specify their dependency chain, at least within a stage and for our own programs, rather than relying on transitive dependencies. That makes it easier to grep for dependencies in the source tree, and easier to reason about the impact of moving or reordering any particular service.

bcressey commented on packages/release/activate-configured.service:
```diff
@@ -0,0 +1,13 @@
+
+[Service]
+Type=oneshot
+ExecStart=/usr/bin/systemctl isolate configured
```
For consistency, should we also set the default?
```suggestion
ExecStart=/usr/bin/systemctl set-default configured
ExecStart=/usr/bin/systemctl isolate configured
```

arnaldo2792 replied:
I could but I wanted to minimize the calls to `set-default` because it will effectively create/remove symlinks.

bcressey replied:
I'd prefer the consistency and the ability to use command-line tools to inspect what state the host system thinks it's reached.

bcressey commented on packages/release/configured.target:
```diff
@@ -1,5 +1,9 @@
 Description=Bottlerocket user and dynamic configuration complete
-After=settings-applier.service
-Requires=settings-applier.service early-boot-config.service
+Before=multi-user.target
```
Is this effectively serving as a sort of documentation? I don't think we strictly need to specify this, unless something is queuing up the `multi-user.target` ahead of time.

I think it'd be more consistent to give `multi-user.target` -> `After=configured.target`, so that none of the targets specify "Before".

arnaldo2792 replied:
Yeah, it was some sort of "documentation", I wanted the target to be very specific about when in the boot process it runs.

bcressey commented on packages/release/preconfigured.target:
```diff
@@ -0,0 +1,10 @@
+Description=Bottlerocket dynamic pre-configuration
+AllowIsolate=yes
+After=basic.target
+Requires=basic.target settings-applier.service
```
Seems weird to specifically list `settings-applier.service` here; we've got that information in the settings applier unit.

arnaldo2792 replied:
I used `settings-applier` since it is the last "leaf" in the dependency tree, all the services required by `preconfigured.target` are effectively "successful" if `settings-applier` exits successfully

bcressey replied:
This goes along with my comment about disliking transitive dependencies: it's hard to trust and reason about `settings-applier` being correct here. What if we add some other service that starts after settings-applier?

I'd prefer that targets mostly just express dependencies on other targets in their own units, and that we fill in the dependency graph in service units instead.

bcressey commented on sources/api/host-containers/src/main.rs:
```diff
@@ -383,7 +393,19 @@ where
         }
-        systemd_unit.enable_and_start()?;
+
+        // Only start the host container if the systemd target is 'multi-user'
```
```suggestion
        // Only start the host container if the systemd target is 'multi-user', otherwise it will start before the system is fully configured.
```

bcressey commented on packages/docker-engine/docker.service:
```diff
@@ -2,8 +2,8 @@
 BindsTo=containerd.service
-After=network-online.target containerd.service configured.target
-Wants=network-online.target configured.target
+After=network-online.target containerd.service
+Wants=network-online.target
```
I'd prefer to keep `configured.target` in `After` / `Wants`, even if it may be overspecified. Does it cause a problem?

(Same with the similar changes for ecs-agent and kubelet.)

arnaldo2792 replied:
It doesn't cause any problems, I can make the changes.

bcressey commented on packages/docker-engine/docker.service:
```diff
@@ -25,3 +25,4 @@ TasksMax=infinity
 
 [Install]
 WantedBy=multi-user.target
+RequiredBy=mark-successful-boot.service
```
We may also want a `Before=mark-successful-boot.service` in the unit section to specify ordering as well as dependency.

(Same with the similar changes for ecs-agent and kubelet.)

arnaldo2792 replied:
Based on our offline conversation, we will only use `RequiredBy=migrator.service` in the `mark-successful-boot.service`

bcressey commented on packages/release/preconfigured.target:
```diff
@@ -0,0 +1,10 @@
+AllowIsolate=yes
+After=basic.target
+Requires=basic.target settings-applier.service
+RefuseManualStart=true
+RefuseManualStop=true
```
I had the sense you said we couldn't add these to the targets or it would break `systemctl isolate`, but it doesn't seem like it does.

Can we add them to the other targets, and potentially all of our required one-shots?

arnaldo2792 replied:
`preconfigured` is queued by systemd itself, so _that_ call isn't manual and using those two configurations here works

I could try to add the configuration to the oneshots, I don't see why they couldn't use these settings, but the other targets can't use them.

bcressey commented on packages/os/settings-applier.service:
```diff
@@ -1,8 +1,7 @@
 [Unit]
 Description=Applies settings to create config files
-After=storewolf.service sundog.service early-boot-config.service apiserver.service
-Requires=apiserver.service
-Wants=storewolf.service sundog.service early-boot-config.service
```
Why remove all of the other dependency information? `sundog` is not the only service that's needed for `settings-applier.service` to succeed.

arnaldo2792 replied:
`sundog` requires `early-boot-config.service`, which itself requires `apiserver.service`. So by specifying just `sundog`, we get the exact same behavior.

bcressey commented:
Re: preconfigured section - `apiserver` is not a one-shot.

One question is whether `systemctl restart apiserver` or a stop & start causes the system to fail catastrophically because all the one-shots in "preconfigured" require it. If so, we might want to relax those dependencies to "Wants" instead.

bcressey commented on packages/chrony/chronyd.service:
```diff
@@ -1,12 +1,12 @@
 Documentation=https://chrony.tuxfamily.org
-After=network-online.target configured.target
-Requires=network-online.target configured.target
+After=network-online.target
```
This should still come after `preconfigured.target`:

```suggestion
After=network-online.target preconfigured.target
```

bcressey commented on packages/os/metricdog-successful-boot.service:
```diff
@@ -0,0 +1,14 @@
+ExecStart=-/usr/bin/metricdog send-boot-success
+
+[Install]
+WantedBy=preconfigured.target
```
If we're sending a ping that includes service health, this should happen in `multi-user.target` when the relevant services have started.

bcressey commented on packages/os/mark-successful-boot.service:
```diff
@@ -1,15 +1,14 @@
 
 [Install]
-WantedBy=multi-user.target
+WantedBy=preconfigured.target
```
If we don't mark the boot as successful, we could potentially run for a very long time with no issues, and then fail back to a previous release on reboot.

```suggestion
RequiredBy=preconfigured.target
```

bcressey commented on packages/os/mark-successful-boot.service:
```diff
@@ -1,15 +1,14 @@
 Description=Call signpost to mark the boot as successful after all required targets are met.
-After=multi-user.target
+After=migrator.service
+Requires=migrator.service
```
```suggestion
Requires=migrator.service
RefuseManualStart=true
RefuseManualStop=true
```

bcressey commented on packages/os/mark-successful-boot.service:
```diff
@@ -1,15 +1,14 @@
+After=migrator.service
+Requires=migrator.service
 # Each service that must start correctly in order for a boot to be successful should be of type "notify"
 # and include "RequiredBy=mark-successful-boot.service" in its [Install] section.
```
We should get rid of this comment.

bcressey commented on packages/os/early-boot-config.service:
```diff
@@ -2,10 +2,15 @@
 # We only want to run once, at first boot.  This file is created by early-boot-config
 # after a successful run.
 ConditionPathExists=!/var/lib/bottlerocket/early-boot-config.ran
+# Prevent manually starting/stoping the service
```
typo:
```suggestion
# Prevent manually starting/stopping the service
```

There are a couple of other cases of this.

bcressey commented on packages/os/send-boot-success.service:
```diff
@@ -0,0 +1,14 @@
+[Unit]
+Description=Send successful boot metrics
```
nit:
```suggestion
Description=Send boot success
```

bcressey commented on packages/os/early-boot-config.service:
```diff
@@ -2,10 +2,15 @@
 # Need network online to talk to IMDS.
 After=network-online.target apiserver.service
-Requires=network-online.target apiserver.service
+Requires=network-online.target
+# We don't want to restart the unit if apiserver restarts
+Wants=apiserver.service
```
`network-online.target` shouldn't be fallible but the same rationale would apply to it:
```suggestion
# We don't want to restart the unit if the network goes offline or apiserver restarts
Wants=network-online.target apiserver.service
```

bcressey commented on packages/os/os.spec:
```diff
@@ -34,6 +34,7 @@ Source107: host-containers@.service
 Source110: mark-successful-boot.service
 Source111: metricdog.service
 Source112: metricdog.timer
+Source113: metricdog-successful-boot.service
```
```suggestion
Source113: send-boot-success.service
```

bcressey commented on packages/os/send-boot-success.service:
```diff
@@ -0,0 +1,14 @@
+[Unit]
+Description=Send successful boot metrics
+# Metricdog depends on the proxy.env file created by settings-applier.service
+After=settings-applier.service
+Requires=settings-applier.service
```
Seems like we can just treat this the same as the other multi-user services:
```suggestion
After=network-online.target configured.target
Wants=network-online.target configured.target
```

bcressey commented on packages/os/sundog.service:
```diff
@@ -1,8 +1,13 @@
+# Needs network access to support commands talking to IMDS.
 After=network-online.target apiserver.service early-boot-config.service
-Requires=network-online.target apiserver.service
+Requires=network-online.target early-boot-config.service
+# We don't want to restart the unit if apiserver restarts
+Wants=apiserver.service
```
```suggestion
Requires=early-boot-config.service
# We don't want to restart the unit if the network goes offline or apiserver restarts
Wants=network-online.target apiserver.service
```

bcressey commented on packages/release/preconfigured.target:
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=Bottlerocket dynamic pre-configuration
```
```suggestion
Description=Bottlerocket initial configuration complete
```

bcressey commented on packages/release/configured.target:
```diff
@@ -1,5 +1,8 @@
 [Unit]
 Description=Bottlerocket user and dynamic configuration complete
```
```suggestion
Description=Bottlerocket final configuration complete
```

bcressey reviewed (commented):
Nice! Just a handful of wording and consistency fixes.

arnaldo2792 commented:
In the last forced pushed I addressed the comments left by @bcressey:

* Typo in `stoping` in multiple units
* Fixed os.spec to include "send-boot-success.service"
* Fixed "send-boot-success.service" dependencies
* Fixed "sundog.service" dependencies
* Fixed "preconfigured.target" and "configured.target" descriptions

bcressey commented on packages/os/early-boot-config.service:
```diff
@@ -2,10 +2,14 @@
 # Need network online to talk to IMDS.
 After=network-online.target apiserver.service
-Requires=network-online.target apiserver.service
+# We don't want to restart the unit the network goes offline or apiserver restarts
```
```suggestion
# We don't want to restart the unit if the network goes offline or apiserver restarts
```

bcressey commented on packages/os/early-boot-config.service:
```diff
@@ -2,10 +2,14 @@
 # We only want to run once, at first boot.  This file is created by early-boot-config
 # after a successful run.
 ConditionPathExists=!/var/lib/bottlerocket/early-boot-config.ran
+# Prevent manually starting/stopping the service
```
We should say "why" rather than "what":

```suggestion
# Block manual interactions with this service, since it could leave the system in an
# unexpected state.
```

arnaldo2792 commented:
In the last forced pushed I addressed the comments left by @bcressey:

* Typo in "early-boot-config.service"
* Improved documentation on why we switched from `Requires` to `Wants` in "early-boot-config.service"
* Improved documentation on why we added the new settings in the "oneshot" units

bcressey commented:
I'm happy with the code changes, but I'm planning to pull down the branch and try some contrived scenarios to see if the system behaves as expected.

arnaldo2792 commented:
* Updated commit message
* Added comments in `mark-successful-boot` and `send-boot-success`

arnaldo2792 commented:
* Rebase upstream
* Add `Requires=storewolf.service` in `early-boot-config.service`

bcressey commented on packages/os/early-boot-config.service:
```diff
@@ -2,10 +2,17 @@
 Description=Bottlerocket userdata configuration system
 # Need network online to talk to IMDS.
 After=network-online.target apiserver.service
```
```suggestion
After=network-online.target apiserver.service storewolf.service
```

arnaldo2792 replied:
Fixed :+1:

arnaldo2792 commented:
* Added `storewolf.service` in `After` for `early-boot-config`

webern commented on sources/api/host-containers/src/main.rs:
```diff
@@ -202,23 +202,30 @@ impl<'a> SystemdUnit<'a> {
             SYSTEMCTL_BIN,
-            &["enable", &self.unit, "--now", "--no-block"],
-        )
+            &["enable", &self.unit, "--now"]
```
Is the removal of --no-block intentional?

bcressey replied:
Before the refactor, this would hold up the boot because of a circular dependency - we couldn't start the container until containerd started, but containerd couldn't start until we were finished applying settings.

I expect it's better to keep `--no-block`, because nothing is waiting on this process and so it may as well exit quickly, and because it could cause trouble later if we start paying more attention to processes spawned by apiserver.

arnaldo2792 replied:
I added `--no-block` to the command in the last forced push I did, as well as the `log-level` flag.

webern reviewed (approved):
`/usr/bin/metricdog send-boot-success` will return zero even if the network is not ready. If that is the case then you would see an error in the journal. If it succeeds there would be no output in the journal unless you run it like this `/usr/bin/metricdog --log-level debug send-boot-success`. Other than making sure send-boot-success works, looks good to me. Nice work!

arnaldo2792 commented:
* Added `--no-block` to the `enable_and_start` command
* Added `--log-level debug` to `send-boot-success`

arnaldo2792 commented:
* @webern  Removed `--log-level debug` from `send-boot-success` since we don't want the entire debug logs here, I verified that `send-boot-success` works.

jahkeup commented on sources/api/host-containers/src/main.rs:
```diff
@@ -229,14 +239,16 @@ where
         .output()
         .context(error::ExecutionFailure { command })?;
 
-    trace!("stdout: {}", String::from_utf8_lossy(&output.stdout));
+    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
+
+    trace!("stdout: {}", stdout);
     trace!("stderr: {}", String::from_utf8_lossy(&output.stderr));
 
     ensure!(
         output.status.success(),
         error::CommandFailure { bin_path, output }
     );
-    Ok(())
+    Ok(stdout)
```
Is there a near-future use case where stdout is handled by callers? If not, then I'd rather this stay as a unit `()` return. The `trace!` can keep printing the value, but I'm not sure it ought be returned to callers based on the changes at call sites in this PR.

arnaldo2792 replied:
I used the `stdout` a little down as part of this change. I use the output to check what's the current default target, and force enable the host containers.

jahkeup replied:
Thanks, I missed that usage!

jahkeup commented on packages/release/activate-configured.service:
```diff
@@ -0,0 +1,14 @@
+[Unit]
```
I'm a bit confused by the use of procedural units to provide what I think are intended to be "hard" synchronization points. systemd provides the means to express the same graph without intermediate `activate-<next>.service` units by way of `<next>.target`s and job ordering. 

To me, the use of `systemctl isolate` seems error prone as unexpressed dependencies may result in a half-working system where a modification of transitive dependency could actually drop a unit from the final set of running units.

This change is somewhat new to me so I might be missing some context - so please forgive me if this was already discussed:

* Did we determine that a `configured-pre.target` and `configured.target` would be unable to express the desired ordering?

* Does `systemctl isolate` provide desired functionality that unit ordering and dependency expression does not? What do we gain by injecting a corrective set of jobs (with `systemctl isolate`) at runtime vs. using dependency ordering?

* Can we avoid replacing the default unit and instead order our units within the default graph?

arnaldo2792 replied:
> Did we determine that a `configured-pre.target` and `configured.target` would be unable to express the desired ordering?

Yes, since the problem that we have is that systemd doesn't start runtime created units (like host containers) unless we force them to start (a.k.a `systemctl enable <> --now`). Force starting units results in units starting/stopping multiple times if any of their dependencies fails to start.

>Does `systemctl isolate` provide desired functionality that unit ordering and dependency expression does not? What do we gain by injecting a corrective set of jobs (with `systemctl isolate`) at runtime vs. using dependency ordering?

Yes, since we can dynamically update the jobs graph at runtime which allows us to add more dependencies to the targets if needed (a.k.a. bootstrap containers). And we gain the ability to dynamically "update" the dependency graph and load the units at the correct time, without forcing starts with `systemctl enable --now`

> Can we avoid replacing the default unit and instead order our units within the default graph?

No, we need `preconfigured` to be the default target for the boot, and transition along the targets to load the systemd services (potentially created during `preconfigured`) without force start them (like we used to do with `host-contaieners`).

jahkeup replied:
Thanks for writing back here. The details will be helpful to maybe-future us scratching our heads reading this. 

We discussed this a bit more offline - this all is roughly what I suspected. I think we should stick with the solution you have implemented here. If we run into issues later on then we can revisit it.

I'm not a fan of `systemctl isolate` by any means, but really this incidental complexity is `systemd`'s fault 😆

jahkeup reviewed (approved):
I'm okay moving forward with this - though I'd like us to use another approach (if that's possible with what systemd provides) if we run into significant issues with `systemctl isolate` down the road.

:shipit:
